From 2cbb408c068b039886eddd126847c6f6a5ee48ca Mon Sep 17 00:00:00 2001
From: P Praneesh <quic_ppranees@quicinc.com>
Date: Wed, 10 Jan 2024 15:07:22 +0530
Subject: [PATCH 1/2] wifi: mac80211: avoid tx pending queue in native wifi
 path

Managing tx pending queue consumes CPU cycles which is not required in
case of performance mode. Skip tx pending queue path and call underlying
driver tx path directly when stats disable flag is set.

Signed-off-by: P Praneesh <quic_ppranees@quicinc.com>
---
 include/net/mac80211.h |  1 +
 net/mac80211/tx.c      | 24 +++++++++++++++++++-----
 2 files changed, 20 insertions(+), 5 deletions(-)

--- a/include/net/mac80211.h
+++ b/include/net/mac80211.h
@@ -3059,6 +3059,7 @@ struct ieee80211_hw {
 	u8 max_txpwr_levels_idx;
 	u32 dbg_mask;
 	u8 medium_busy;
+	bool perf_mode;
 };
 
 static inline bool _ieee80211_hw_check(struct ieee80211_hw *hw,
--- a/net/mac80211/tx.c
+++ b/net/mac80211/tx.c
@@ -3721,9 +3721,12 @@ void __ieee80211_xmit_fast(struct ieee80
 	struct ieee80211_hdr *hdr = (void *)fast_tx->hdr;
 	struct ieee80211_tx_info *info;
 	struct ieee80211_tx_data tx;
+	struct ieee80211_tx_control control = {};
 	ieee80211_tx_result r;
 	int hw_headroom = sdata->local->hw.extra_tx_headroom;
 	int extra_head = fast_tx->hdr_len - (ETH_HLEN - 2);
+	bool skip_tx_queue = (local->hw.perf_mode &&
+			      !ieee80211_vif_is_mesh(&sdata->vif));
 
 	skb = skb_share_check(skb, GFP_ATOMIC);
 	if (unlikely(!skb))
@@ -3777,8 +3780,11 @@ void __ieee80211_xmit_fast(struct ieee80
 	tx.sta = sta;
 	tx.key = fast_tx->key;
 
-	if (ieee80211_queue_skb(local, sdata, sta, skb))
-		return;
+	if (unlikely(!skip_tx_queue)) {
+		__skb_queue_head_init(&tx.skbs);
+		if (ieee80211_queue_skb(local, sdata, sta, skb))
+			return;
+	}
 
 	tx.skb = skb;
 	r = ieee80211_xmit_fast_finish(sdata, sta, fast_tx->pn_offs,
@@ -3791,8 +3797,15 @@ void __ieee80211_xmit_fast(struct ieee80
 		sdata = container_of(sdata->bss,
 				     struct ieee80211_sub_if_data, u.ap);
 
-	__skb_queue_tail(&tx.skbs, skb);
-	ieee80211_tx_frags(local, &sdata->vif, sta, &tx.skbs, false);
+	if (likely(skip_tx_queue)) {
+		info->control.vif = &sdata->vif;
+		control.sta = sta ? &sta->sta : NULL;
+		drv_tx(local, &control, skb);
+	} else {
+		__skb_queue_tail(&tx.skbs, skb);
+		ieee80211_tx_frags(local, &sdata->vif, sta, &tx.skbs, false);
+	}
+
 	return;
 
 free:
@@ -4908,10 +4921,11 @@ netdev_tx_t ieee80211_subif_start_xmit_8
 	struct ieee80211_tx_control control = {};
 	struct sta_info *sta;
 	struct ieee80211_sta *pubsta = NULL;
+	bool perf_mode = sdata->local->hw.perf_mode;
 
 	info->control.vif = &sdata->vif;
 
-	if (skb->fast_xmit) {
+	if (likely(skb->fast_xmit && perf_mode)) {
 		info->control.flags = u32_encode_bits(IEEE80211_LINK_UNSPECIFIED,
 						      IEEE80211_TX_CTRL_MLO_LINK);
 		info->flags = IEEE80211_TX_CTL_HW_80211_ENCAP;
